File: 2stWeek/ddonydev/4.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if(image[sr][sc] == color) {
            return image;
        }

        int oldColor = image[sr][sc];
        dfs(image, sr, sc, oldColor, color);
        return image;
    }
    
    void dfs(int[][] image, int sr, int sc, int oldColor, int color){

        if(sr < 0 || sc < 0 || sr >= image.length || sc >= image[0].length || image[sr][sc] != oldColor) {
            return;
        }

        image[sr][sc] = color;
        dfs(image, sr - 1, sc, oldColor, color);
        dfs(image, sr + 1, sc, oldColor, color);
        dfs(image, sr, sc - 1, oldColor, color);
        dfs(image, sr, sc + 1, oldColor, color);
    }
}

File: 2stWeek/pirit.jeong/4.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function (image, sr, sc, color) {
  const directions = [
    [0, -1],
    [0, 1],
    [-1, 0],
    [1, 0],
  ];

  const row = image.length;
  const col = image[0].length;
  const visited = new Array(row)
    .fill(false)
    .map(() => new Array(col).fill(false));

  recurse(image, sr, sc, image[sr][sc], color, visited);

  function recurse(image, sr, sc, origin, color, visited) {
    if (image[sr][sc] !== origin) {
      return;
    }

    image[sr][sc] = color;
    visited[sr][sc] = true;

    for (const direction of directions) {
      const [dx, dy] = direction;

      const nx = sr + dx;
      const ny = sc + dy;

      if (nx < 0 || nx >= image.length || ny < 0 || ny >= image[0].length) {
        continue;
      } else if (visited[nx][ny]) {
        continue;
      }

      recurse(image, nx, ny, origin, color, visited);
    }
  }

  return image;
};
